core/types/transaction.go

```diff
@@ -415,9 +415,7 @@ func (t *TransactionsByPriceAndNonce) Peek() *Transaction {
 }
 
 // Shift replaces the current best head with the next one from the same account.
-func (t *TransactionsByPriceAndNonce) Shift() {
-	signer := deriveSigner(t.heads[0].data.V)
-	// derive signer but don't cache.
+func (t *TransactionsByPriceAndNonce) Shift(signer Signer) {
```

Comment:
I pinged @karalabe in a voice call and he doesn't like this change because calling `Shift` with some signer other than EIP155 can derive a different sender address and mess up the tx set. A suggestion: maybe we should store the signer to use on the `TransactionsByPriceAndNonce` struct.